File: apps/frontend/lib/api/ats.ts
/**
 * ATS screening API client and TypeScript types.
 */

import { apiPost } from './client';
import type { ResumeData } from '@/components/dashboard/resume-component';

export interface ScoreBreakdown {
  skills_match: number;
  experience_match: number;
  domain_match: number;
  tools_match: number;
  achievement_match: number;
  total: number;
}

export interface KeywordRow {
  keyword: string;
  found_in_resume: boolean;
  section: string | null;
}

export type ATSDecision = 'PASS' | 'BORDERLINE' | 'REJECT';

export interface ATSScreeningResult {
  score: ScoreBreakdown;
  decision: ATSDecision;
  keyword_table: KeywordRow[];
  missing_keywords: string[];
  warning_flags: string[];
  optimization_suggestions: string[];
  optimized_resume: ResumeData | null;
  saved_resume_id: string | null;
}

export interface ATSScreenRequest {
  resume_id?: string;
  resume_text?: string;
  job_id?: string;
  job_description?: string;
  save_optimized?: boolean;
}

export async function screenResume(
  request: ATSScreenRequest
): Promise<ATSScreeningResult> {
  const resp = await apiPost('/ats/screen', request, 120_000);
  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: 'ATS screening failed' }));
    throw new Error(err.detail ?? 'ATS screening failed');
  }
  return resp.json();
}

export interface ATSSaveResumeRequest {
  resume_data: ResumeData;
  parent_id?: string | null;
  title?: string;
}

export interface ATSSaveResumeResponse {
  resume_id: string;
}

export async function saveAtsResume(
  request: ATSSaveResumeRequest
): Promise<ATSSaveResumeResponse> {
  const resp = await apiPost('/ats/save-resume', request, 30_000);
  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: 'Save failed' }));
    throw new Error(err.detail ?? 'Save failed');
  }
  return resp.json();
}

File: apps/backend/app/schemas/ats.py
"""Pydantic schemas for ATS screening."""

from typing import Literal

from pydantic import BaseModel

from app.schemas.models import ResumeData


class ATSScreenRequest(BaseModel):
    """Request body for POST /api/v1/ats/screen."""

    resume_id: str | None = None
    resume_text: str | None = None
    job_id: str | None = None
    job_description: str | None = None
    save_optimized: bool = False


class ScoreBreakdown(BaseModel):
    """Weighted ATS score per dimension."""

    skills_match: float      # max 30
    experience_match: float  # max 25
    domain_match: float      # max 20
    tools_match: float       # max 15
    achievement_match: float # max 10
    total: float             # 0-100


class KeywordRow(BaseModel):
    """One row in the keyword match table."""

    keyword: str
    found_in_resume: bool
    section: str | None = None  # e.g. "workExperience", "summary", null if not found


class ATSScreeningResult(BaseModel):
    """Full ATS screening report returned by the endpoint."""

    score: ScoreBreakdown
    decision: Literal["PASS", "BORDERLINE", "REJECT"]
    keyword_table: list[KeywordRow]
    missing_keywords: list[str]
    warning_flags: list[str]           # >= 10 items when decision == "REJECT"
    optimization_suggestions: list[str]
    optimized_resume: ResumeData | None = None
    saved_resume_id: str | None = None  # populated when save_optimized=True


class ATSSaveResumeRequest(BaseModel):
    """Request body for POST /api/v1/ats/save-resume."""

    resume_data: ResumeData
    parent_id: str | None = None
    title: str = "ATS Optimized Resume"


class ATSSaveResumeResponse(BaseModel):
    """Response body for POST /api/v1/ats/save-resume."""

    resume_id: str
